### PR TITLE
feat: tyre wear extrapolator — weather segmentation, sliding window, SC-gap resilience

### PR DESCRIPTION
**Pull request by PettableCat**
## 📝 Description

Overhauls the tyre wear extrapolator to handle mid-race weather transitions, safety car gaps, and short stints more robustly.

Key changes:
- Weather-aware segmentation: splits tyre wear history into dry/wet segments so predictions don't blend incompatible data
- Sliding-window regression replaces full-stint regression for more responsive wear curves
- Safety car gap resilience: filters out laps with anomalous low wear caused by SC periods
- Adds `isDry()` / `isWet()` helpers to `WeatherCondition` enum (replaces raw int comparisons)
- Moves `_MIN_WEATHER_SEGMENT_LAPS` class variable above `__init__` for readability

---

## 📂 Type of change

- [ ] Bug fix 🐞
- [x] New feature 🚀
- [x] Refactor 🔨
- [ ] Documentation 📚
- [x] Tests ✅
- [ ] Other:

---

### ✅ Integration Tests

* [x] All integration tests pass
* Attach test command/output:

```
$ poetry run python tests/integration_test/runner.py

Files processed:        31
Telemetry sent:         31
Telemetry failed:       0
Endpoint checks passed: 806
Endpoint checks failed: 0
Telemetry success rate: 100.0%
Endpoint success rate:  100.0%
Total execution time: 44:03.297
```

To run the integration test
```bash
poetry run python tests/integration_test/runner.py
```

---

## 🧱 `lib/` Directory Changes

* [x] This PR modifies or adds files under `lib/`
* [x] I have updated or added unit tests to reflect those changes

Explain your test updates here:

```
- lib/f1_types/packet_1_session_data.py: Added isDry()/isWet() helpers to WeatherCondition enum
- lib/tyre_wear_extrapolator/tyre_wear_extrapolator.py: Weather segmentation, sliding window, SC-gap filtering
- lib/tyre_wear_extrapolator/tyre_wear_per_lap.py: Updated to use WeatherCondition enum
- tests/tests_tyre_wear_extrapolator.py: 55 tests covering weather transitions, SC gaps, short stints, edge cases
```

---

## 📋 General Checklist

* [x] I have read the [CONTRIBUTING](../CONTRIBUTING.md) guide
* [x] My code follows project style conventions
* [x] All new and existing tests pass
* [x] I have added relevant documentation/comments
* [x] I have reviewed my changes and believe they are ready to merge

---

### CODE REVIEW

File: lib/tyre_wear_extrapolator/tyre_wear_extrapolator.py [hunk 1]
```diff
@@ -271,18 +277,68 @@ def _updateDataList(self, new_data: List[TyreWearPerLap]):
         self._recomputeRacingLapsData()
         self._recompute()
 
+    _MIN_WEATHER_SEGMENT_LAPS = 3
```
Comment (ashwin-nat) [inline lib/tyre_wear_extrapolator/tyre_wear_extrapolator.py]: please specify all class vars at the top of the class (above ctor)

File: lib/tyre_wear_extrapolator/tyre_wear_extrapolator.py [hunk 2]
```diff
@@ -384,57 +441,92 @@ def _extrapolateTyreWear(self) -> None:
...
                     rl_tyre_wear=rl_wear,
                     rr_tyre_wear=rr_wear,
-                    lap_number=lap,
+                    lap_number=actual_lap,
                 )
                 self.m_predicted_tyre_wear.append(predicted_tyre)
         self._enforce_monotonicity()
 
+    # Weather grouping: Dry (Clear=0, LightCloud=1, Overcast=2) vs Wet (LightRain=3, HeavyRain=4, Storm=5, Thunderstorm=6)
+    _DRY_WEATHER_IDS = frozenset({0, 1, 2})
```
Comment (ashwin-nat) [inline lib/tyre_wear_extrapolator/tyre_wear_extrapolator.py]: well defined types for weather is already available in WeatherForecastSample.WeatherCondition. please use that everywhere instead of int to improve readability and debuggability. int is very opaque and if the spec changes in F1 27, this will be hard to catch.

You can even add a helper method in WeatherForecastSample.WeatherCondition called isDry and isWet. similar to isRaceTypeSession in SessionType

File: tests/tests_tyre_wear_extrapolator.py [hunk 3]
```diff
@@ -301,3 +301,488 @@ def test_recovery_after_missing_laps(self):
...
+        # But total racing laps collected is still 5
+        self.assertEqual(extrap.num_samples, 5)
+
+
+class TestTyreWearExtrapolatorWeatherSegmentation(TestTyreWearPrediction):
+    """Test weather-aware segmentation: dry/wet grouping, fallback logic,
+    and backward compatibility with legacy data (weather_id=None)."""
+
+    # Weather enum constants
+    CLEAR = 0
```
Comment (ashwin-nat) [inline tests/tests_tyre_wear_extrapolator.py]: same comment as above, reuse the existing enums. it can even be serialised to string for readability. and we get the same performance gains as using int

**Comment by ashwin-nat** [comment]
what happens if say the first sample will not have weather info? this may happen in 2 scenarios.

1. the zeroth lap snapshot. there is no guarantee that the session packet may have arrived before the other required data
2. if the app is started late and a snapshot is being recorded, the first data point may not have weather info

These are both valid scenarios

**Comment by PettableCat** [comment]
Both scenarios are handled. `weather_id` on `TyreWearPerLap` defaults to `None` (see `tyre_wear_per_lap.py:50`), and the code treats `None` as "unknown" throughout:

1. **`_weather_group(None)` returns `None`** — so no weather group is assigned.
2. **`_filter_to_current_weather()`** — if the most recent lap has `weather_id=None`, it returns all data unchanged (line ~293: `if current_group is None: return racing_data`).
3. **`_segmentData()`** — a `None` weather group never forces a segment break (line ~516: `point_weather_group is not None and weather_group is not None` — both must be non-None for a weather change to be detected).

So if the app starts late or the session packet hasn't arrived yet, the first samples simply carry `weather_id=None` and the extrapolator falls back to using all available data — identical to the behavior before this PR. Once weather data starts flowing, subsequent laps carry the actual weather enum and segmentation kicks in from that point forward.

This is also covered by `test_all_none_weather_backward_compat` which verifies that data without any weather info produces the same regression results as before.
